cli/clistat/cgroup.go

```diff
@@ -0,0 +1,296 @@
+	usageUs, err := readInt64Prefix(s.fs, cgroupV2CPUStat, "usage_usec")
+	if err != nil {
+		return 0, xerrors.Errorf("get cgroupv2 cpu used: %w", err)
+	}
+	periodUs, err := readInt64SepIdx(s.fs, cgroupV2CPUMax, " ", 1)
+	if err != nil {
+		return 0, xerrors.Errorf("get cpu period: %w", err)
+	}
+
+	return float64(usageUs) / float64(periodUs), nil
+}
+
+func (s *Statter) cGroupV2CPUTotal() (total float64, err error) {
+	var quotaUs, periodUs int64
+	periodUs, err = readInt64SepIdx(s.fs, cgroupV2CPUMax, " ", 1)
+	if err != nil {
+		return 0, xerrors.Errorf("get cpu period: %w", err)
+	}
+
+	quotaUs, err = readInt64SepIdx(s.fs, cgroupV2CPUMax, " ", 0)
+	if err != nil {
+		// Fall back to number of cores
+		quotaUs = int64(s.nproc) * periodUs
+	}
+
+	return float64(quotaUs) / float64(periodUs), nil
+}
+
+func (s *Statter) cGroupV1CPUTotal() (float64, error) {
+	periodUs, err := readInt64(s.fs, cgroupV1CFSPeriodUs)
+	if err != nil {
+		return 0, xerrors.Errorf("read cpu period: %w", err)
+	}
+
+	quotaUs, err := readInt64(s.fs, cgroupV1CFSQuotaUs)
+	if err != nil {
+		return 0, xerrors.Errorf("read cpu quota: %w", err)
+	}
+
+	if quotaUs < 0 {
+		// Fall back to the number of cores
+		quotaUs = int64(s.nproc) * periodUs
+	}
+
+	return float64(quotaUs) / float64(periodUs), nil
+}
+
+func (s *Statter) cGroupV1CPUUsed() (float64, error) {
+	usageNs, err := readInt64(s.fs, cgroupV1CPUAcctUsage)
+	if err != nil {
+		// try alternate path
+		usageNs, err = readInt64(s.fs, cgroupV1CPUAcctUsageAlt)
+		if err != nil {
+			return 0, xerrors.Errorf("read cpu used: %w", err)
+		}
+	}
+
+	// usage is in ns, convert to us
+	usageNs /= 1000
+	periodUs, err := readInt64(s.fs, cgroupV1CFSPeriodUs)
+	if err != nil {
+		return 0, xerrors.Errorf("get cpu period: %w", err)
+	}
+
+	return float64(usageNs) / float64(periodUs), nil
+}
+
+// ContainerMemory returns the memory usage of the container cgroup.
+// If the system is not containerized, this always returns nil.
+func (s *Statter) ContainerMemory() (*Result, error) {
+	if ok, err := IsContainerized(s.fs); err != nil || !ok {
+		return nil, nil //nolint:nilnil
+	}
+
+	if s.isCGroupV2() {
+		return s.cGroupV2Memory()
+	}
+
+	// Fall back to CGroupv1
+	return s.cGroupV1Memory()
+}
+
+func (s *Statter) cGroupV2Memory() (*Result, error) {
+	maxUsageBytes, err := readInt64(s.fs, cgroupV2MemoryMaxBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("read memory total: %w", err)
+	}
+
+	currUsageBytes, err := readInt64(s.fs, cgroupV2MemoryUsageBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("read memory usage: %w", err)
+	}
+
+	inactiveFileBytes, err := readInt64Prefix(s.fs, cgroupV2MemoryStat, "inactive_file")
+	if err != nil {
+		return nil, xerrors.Errorf("read memory stats: %w", err)
+	}
+
+	return &Result{
+		Total: ptr.To(float64(maxUsageBytes) / 1024 / 1024 / 1024),
+		Used:  float64(currUsageBytes-inactiveFileBytes) / 1024 / 1024 / 1024,
+		Unit:  "GB",
```

Comment:
Loose thought, but wouldn't like to block you.

"Result" stores a statistic with the approximate format. If somebody would like to use it to check a condition `<100 MB`, they will have to look up the format, and then possibly convert. I'm wondering if it isn't simple to just keep it as basic units (bytes, seconds, percentage, etc.) and transform it to "human readable" when rendering the output. 

+ mixing responsibilities here: value store vs formatting.

Reply:
That's fair.